File: counting-bits.cpp
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> countBits(int n) {
    vector<int> counts(n+1, 0);

    for (int i=1; i<=n; ++i) {
      counts[i] = counts[i >> 1] + (i & 1);
    }

    return counts;
  }
};

File: counting-bits.py
from typing import List

 class Solution:
    def countBits(self, n: int) -> List[int]:
        counts = [0] * (n+1)
        
        for i in range(1, n+1):
            counts[i] = counts[i >> 1] + (i & 1)

        return counts
